### PR TITLE
Deep linking (closes #331)

### LINKED ISSUE #331: Deep linking to POI View (android)
When a notification is tapped the app should open directly into the corresponding POI view.

### PR DESCRIPTION
**Pull request by wilsonpage**

### CODE REVIEW

**Comment by wilsonpage** [comment]
@arcturus r? (ping-pong sorry)

File: lib/store/constants.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+module.exports = {
+  INCOMPLETE: 'incomplete',
```
Comment (wilsonpage) [inline lib/store/constants.js:3]: `FETCHING`

File: lib/store/reducer.js [hunk 2]
```diff
@@ -4,18 +4,24 @@
...
+  ERROR,
+} = require('./constants');
+
 const initialState = {
-  items: [],
-  openItem: {
-    isFetching: false,
-    value: null,
-  },
+  items: {},
```
Comment (wilsonpage) [inline lib/store/reducer.js:18]: `itemsCache`?

File: lib/views/item/item.js [hunk 3]
```diff
@@ -2,8 +2,8 @@
  * Dependencies
```
Comment (wilsonpage) [inline lib/views/item/item.js]: Change filename to `item-detail.js`

File: lib/views/item/container.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
+'use strict';
```
Comment (wilsonpage) [inline lib/views/item/container.js]: -> `item-container.js`

File: lib/views/list/item.js [hunk 5]
```diff
@@ -5,8 +5,8 @@
  */
```
Comment (wilsonpage) [inline lib/views/list/item.js]: `list-item.js`

File: lib/views/item/scene.js [hunk 6]
```diff
@@ -4,28 +4,42 @@
...
-  navigator: React.PropTypes.object.isRequired,
-  itemUrl: React.PropTypes.string,
-  item: React.PropTypes.object,
-  setOpenItem: React.PropTypes.func,
+  navigator: PropTypes.object.isRequired,
+  itemUrl: PropTypes.string,
+  openItem: PropTypes.string,
+  items: PropTypes.object,
+  setOpenItem: PropTypes.func,
+  dispatch: PropTypes.func.isRequired,
```
Comment (wilsonpage) [inline lib/views/item/scene.js]: Maybe can remove some of these?

File: lib/views/list/scene.js [hunk 7]
```diff
@@ -139,30 +159,31 @@ ListScene.propTypes = {
  * @param  {ReduxStore} store
  * @return {Object}
  */
-function mapStateToProps(state) {
-  const { sortByDistance, showDistance } = state.userFlags;
-  const items = state.itemsNearby
-    .map(url => resolveItem(url))
-    .filter(item => !!item);
+function mapStateToProps({ items, itemsNearby, userFlags, indicateScanning }) {
```
Comment (wilsonpage) [inline lib/views/list/scene.js]: Remove `items`

Comment (wilsonpage) [reply lib/views/list/scene.js]: Ignore

File: ios/ApiMetadata.swift [hunk 8]
```diff
@@ -35,20 +35,25 @@ class ApiMetadata: ApiBase {
   //
   //  data["objects"] = elems
```
Comment (wilsonpage) [inline ios/ApiMetadata.swift:36]: Remove comment

Comment (wilsonpage) [reply ios/ApiMetadata.swift:36]: *Update

**Comment by arcturus** [comment]
Follow ups:
- Make a service that returns both metadata + beacon data. Outside of metadata server, remove the api in js land that do the mixing (https://github.com/mozilla-magnet/magnet-services/issues/30)
- Batch requests about objects (https://github.com/mozilla-magnet/magnet-client/issues/347).
- onLost logic for ios scanners. (https://github.com/mozilla-magnet/magnet-scanner-ios/issues/13)
- Cache of items, at some point clear old items, but when the items get complicated. Now it's not a problem (https://github.com/mozilla-magnet/magnet-client/issues/346)
- Rename files to match class name.
- Calculate distance on the client from strength or lat lon, but we won't update the model that often, on the native side we should send a power mashed up (https://github.com/mozilla-magnet/magnet-scanner-ios/issues/12, https://github.com/mozilla-magnet/magnet-scanner-android/issues/25)
- When launching the app directly from deep linking dont show the list of items (https://github.com/mozilla-magnet/magnet-client/issues/348)

**Comment by arcturus** [comment]
@wilsonpage after rebasing to latest master (it has a patch from @samgiles for react-native version) it does kind of work in iphone.

It at least tryies to load the view, it loads the basic, and then the app dies cause an exception in the new postArray method, need to debug more to understand whats going on.

**Comment by arcturus** [comment]
More feedback about what's going on.

It's important to take into account the environment where I am, surronded by beacons that we constantly receive, so I have hundreds of calls to app render scene home in a matter of 10 seconds. Either that or we are constantly triggering model change, and also calling fetchItemIfNeeded even if it's going to end soon, we still have an enormous number of calls to that function.

That is making the animation to open the item to be very slow and to compite for resources. When we load an item that is in the cache everything seems to work ok.

**Comment by arcturus** [comment]
Yes, I can confirm that in the simulator without so many update cause of btle beacons everything works way smoother.

We can consider this like an aggressive environment, but we should think on adding some throttling to that update coming from BLE.

**Comment by wilsonpage** [comment]
@arcturus:
- Fixed perf issue caused by transitions being hammered by react re-render
- Tested deep links working on real ios device
- Ready to land

**Comment by arcturus** [comment]
Shiiiiip it once travis is green :)

**Pull request merged**